### PR TITLE
Handling Medtronic temp basal edge cases

### PR DESCRIPTION
**Pull request by gniezen**
The binary blob provided by Darin uncovered a number of edges related to temp basals (probably because his pump is running OpenAPS):

- temp basal without a duration record (a temp basal is usually followed by a duration record)
- a new record type 0x20
- temp basal set to zero percent (instead of zero rate)
- a running temp basal with specified rate updated with new percentage (instead of new rate)

### CODE REVIEW

**Comment by gniezen** [comment]
[[uploader] Medtronic basals](https://trello.com/c/zlAnS0iw/13-uploader-medtronic-basals)

File: lib/medtronic/medtronicSimulator.js [hunk 1]
```diff
@@ -246,26 +246,34 @@ exports.make = function(config){
...
           if (event.duration === 0) {
-            // temp basal was cancelled:
-            // The pump sends a temp basal record with duration 0.
-            // We use the time this record was sent to calculate the actual duration.
-            checkForScheduleChanges(event);
-            currBasal.with_expectedDuration(currBasal.duration);
-            var duration = Date.parse(event.time) - Date.parse(currBasal.time);
-            currBasal.with_duration(duration);
+            if(event.annotations && event.annotations[0].code === 'basal/unknown-duration') {
+              console.log('Temp basal at ' + event.time + ' has unknown duration');
```
Comment (pazaan) [inline lib/medtronic/medtronicSimulator.js]: debug instead of console.log?

File: lib/medtronic/processData.js [hunk 2]
```diff
@@ -902,17 +903,19 @@ var buildTempBasalRecords = function (records) {
         .set('index', record.index)
         .set('jsDate',record.jsDate);
 
-      if(percentage) {
+      if(percentage != null) {
```
Comment (pazaan) [inline lib/medtronic/processData.js]: Should we be using the !== style?

File: lib/medtronic/processData.js [hunk 3]
```diff
@@ -902,17 +903,19 @@ var buildTempBasalRecords = function (records) {
...
-      if(durationRecord && durationRecord.type.value === RECORD_TYPES.TEMP_BASAL_DURATION.value) {
+      var durationRecord = basalRecords[i+1];
+      if(durationRecord && durationRecord.type.value === RECORD_TYPES.TEMP_BASAL_DURATION.value &&
+          durationRecord.jsDate.valueOf() === tempBasal.jsDate.valueOf()) {
+        i += 1; // advance past duration record
         tempBasal = tempBasal.with_duration(struct.extractByte(durationRecord.head,1) * 30 * sundial.MIN_TO_MSEC);
       } else {
-        console.log('Records not matching:',record, durationRecord);
-        throw new Error('Missing record');
+        console.log('Temp basal not followed by duration:',record, durationRecord);
```
Comment (pazaan) [inline lib/medtronic/processData.js]: `debug`?

File: lib/medtronic/processData.js (hunk 2 above)
Comment (gniezen) [inline lib/medtronic/processData.js]: In a lot of the driver code we use `!= null` as a check for both `null` and `undefined`.

Comment (pazaan) [inline lib/medtronic/processData.js]: 👍

**Pull request merged**